src/lambda/wizards/samDeployWizard.ts

```diff
@@ -524,11 +487,75 @@ export class DefaultSamDeployWizardContext implements SamDeployWizardContext {
         })
     }
 
+    private async addBucketsToRegionMap(buckets: S3.Bucket[], s3Client: S3Client): Promise<void> {
+        const promises = buckets
+            .filter(bucket => bucket.Name)
+            .map(async bucket => {
+                const bucketLocation = await s3Client.getBucketLocation(bucket.Name!)
+                const transformedLocation = this.normalizeLocation(bucketLocation.LocationConstraint)
+                this.addBucketToRegionMap(bucket.Name!, transformedLocation)
+            })
+
+        await Promise.all(promises)
+    }
+
+    // correct blank and 'EU' locations:
+    // https://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region
+    private normalizeLocation(location?: string): string {
+        if (!location || location === '') {
+            location = 'us-east-1'
+        } else if (location === 'EU') {
+            location = 'eu-west-1'
+        }
+
+        return location
+    }
+
     private addBucketToRegionMap(bucket: string, region: string) {
-        if (!this.regionBucketMap.has(region)) {
-            this.regionBucketMap.set(region, [])
+        if (!this.bucketsFilteredByRegion.has(region)) {
+            this.bucketsFilteredByRegion.set(region, [])
+        }
+        this.bucketsFilteredByRegion.get(region)!.push(bucket)
+    }
+
+    private createBucketQuickPickItems(
+        selectedRegion: string,
+        noBucketsMessage: string,
+        initialValue?: string
+    ): vscode.QuickPickItem[] {
+        let buckets: vscode.QuickPickItem[] | undefined
+
+        // does the selected region have buckets? If so, add to the QP
+        if (this.bucketsFilteredByRegion.size > 0) {
+            const regionBucket = this.bucketsFilteredByRegion.get(selectedRegion)
+            if (regionBucket && regionBucket.length > 0) {
+                // turn these into quick pick items
+                buckets = regionBucket.map(b => (
+                    {
+                        label: b!,
+                        // this is the only way to get this to show on going back
+                        // this will make it so it always shows even when searching for something else
+                        alwaysShow: b === initialValue,
+                        description: b === initialValue ?
+                            localize('AWS.samcli.wizard.selectedPreviously', 'Selected Previously') : ''
+                    }
+                ))
+            }
+        }
+        // add the buckets to the QP OR add a message that buckets don't exist (and prompt to go back)
+        if (buckets && buckets.length > 0) {
```

Comment:
This logic can be simplified if you unconditionally create the array on line 526.

Then this is simply...
```
IF buckets is empty THEN
   ADD FALLBACK ITEM
ENDIF

RETURN buckets
```

Reply:
Will add on next revision

Reply:
✔️